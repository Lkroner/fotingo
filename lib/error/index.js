'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.errors = exports.catchPromiseAndThrow = exports.handleError = exports.throwControlledError = exports.ControlledError = undefined;

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

var _errors = require('./errors');

var _errors2 = _interopRequireDefault(_errors);

var _util = require('../util');

var _reporter = require('../reporter');

var _reporter2 = _interopRequireDefault(_reporter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _extendableBuiltin(cls) {
  function ExtendableBuiltin() {
    var instance = Reflect.construct(cls, Array.from(arguments));
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    return instance;
  }

  ExtendableBuiltin.prototype = Object.create(cls.prototype, {
    constructor: {
      value: cls,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });

  if (Object.setPrototypeOf) {
    Object.setPrototypeOf(ExtendableBuiltin, cls);
  } else {
    ExtendableBuiltin.__proto__ = cls;
  }

  return ExtendableBuiltin;
}

var ControlledError = exports.ControlledError = function (_extendableBuiltin2) {
  _inherits(ControlledError, _extendableBuiltin2);

  function ControlledError(message) {
    var parameters = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, ControlledError);

    return _possibleConstructorReturn(this, (ControlledError.__proto__ || Object.getPrototypeOf(ControlledError)).call(this, _ramda2['default'].compose(_ramda2['default'].reduce(function (msg, _ref) {
      var _ref2 = _slicedToArray(_ref, 2),
          k = _ref2[0],
          v = _ref2[1];

      return _ramda2['default'].replace('${' + k + '}', v, msg);
    }, message), _ramda2['default'].toPairs)(parameters)));
  }

  return ControlledError;
}(_extendableBuiltin(Error));

var throwControlledError = exports.throwControlledError = function () {
  function throwControlledError(message, parameters) {
    return function () {
      throw new ControlledError(message, parameters);
    };
  }

  return throwControlledError;
}();

// Error -> Boolean
var isKnownError = _ramda2['default'].either(_ramda2['default'].is(ControlledError), _ramda2['default'].propEq('message', 'canceled'));
var exit = function () {
  function exit(code) {
    return function () {
      return process.exit(code);
    };
  }

  return exit;
}();
var userIsExiting = _ramda2['default'].compose(_ramda2['default'].equals('canceled'), _ramda2['default'].prop('message'));
var handleErrorAndExit = _ramda2['default'].compose(exit(0), _util.error, _ramda2['default'].prop('message'));
var handleUnknownError = _ramda2['default'].compose(exit(1), _util.error);
var sayBye = function () {
  function sayBye() {
    return _reporter2['default'].log('Hasta la vista baby!', 'wave');
  }

  return sayBye;
}();
var handleError = exports.handleError = _ramda2['default'].ifElse(isKnownError, _ramda2['default'].ifElse(userIsExiting, sayBye, handleErrorAndExit), handleUnknownError);
// String -> Promise -> Promise
var catchPromiseAndThrow = exports.catchPromiseAndThrow = function () {
  function catchPromiseAndThrow(module, e) {
    return function (p) {
      return p['catch'](function (err) {
        if (_ramda2['default'].is(Function, e)) {
          throwControlledError(e(err))(err);
        } else {
          _ramda2['default'].compose(throwControlledError(e), (0, _util.debugCurried)(module, err))(err);
        }
      });
    };
  }

  return catchPromiseAndThrow;
}();
exports.errors = _errors2['default'];