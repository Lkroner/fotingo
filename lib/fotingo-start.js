'use strict';

var _commander = require('commander');

var _commander2 = _interopRequireDefault(_commander);

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

var _util = require('./issue-tracker/util');

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

var _error = require('./error');

var _init = require('./init');

var _init2 = _interopRequireDefault(_init);

var _reporter = require('./reporter');

var _reporter2 = _interopRequireDefault(_reporter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _reporter$stepFactory = _reporter2['default'].stepFactory(4),
    step = _reporter$stepFactory.step,
    stepCurried = _reporter$stepFactory.stepCurried;

var getIssueId = _ramda2['default'].compose(_util.validate, _ramda2['default'].head, _ramda2['default'].prop('args'));

try {
  (function () {
    _commander2['default'].parse(process.argv);
    var issueId = getIssueId(_commander2['default']);
    step(1, 'Initializing services', 'rocket');
    (0, _init2['default'])(_config2['default'], _commander2['default']).then(function (_ref) {
      var git = _ref.git,
          issueTracker = _ref.issueTracker;
      return issueTracker.getCurrentUser().then(stepCurried(2, 'Getting issue from ' + issueTracker.name, 'bug')).then(function (user) {
        return issueTracker.getIssue(issueId).then(issueTracker.canWorkOnIssue(user)).then(stepCurried(3, 'Updating issue status', 'bookmark')).then(issueTracker.setIssueStatus({ status: issueTracker.status.IN_PROGRESS })).then(stepCurried(4, 'Creating branch to work on issue', 'tada')).then(git.createIssueBranch(_config2['default']));
      });
    }).then(_reporter2['default'].footer)['catch'](_error.handleError);
  })();
} catch (e) {
  (0, _error.handleError)(e);
  _commander2['default'].help();
}