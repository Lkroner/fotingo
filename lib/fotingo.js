#!/usr/bin/env node
'use strict';

var _commander = require('commander');

var _commander2 = _interopRequireDefault(_commander);

var _package = require('../package.json');

var _package2 = _interopRequireDefault(_package);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

_commander2['default'].version(_package2['default'].version).command('start [issue-id]', 'start working in a new issue').command('review', 'submit current issue for review').parse(process.argv);