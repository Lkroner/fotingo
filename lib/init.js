'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

var _local = require('./git/local');

var _local2 = _interopRequireDefault(_local);

var _github = require('./git/github');

var _github2 = _interopRequireDefault(_github);

var _issueTracker = require('./issue-tracker');

var _issueTracker2 = _interopRequireDefault(_issueTracker);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

// config -> project -> promise
exports['default'] = _ramda2['default'].curryN(2, function (c, program) {
  return _github2['default'].init(c)().then(_local2['default'].init(c, process.cwd())).then((0, _issueTracker2['default'])(program)(c)).then(function (issueTracker) {
    return { github: _github2['default'], git: _local2['default'], issueTracker: issueTracker };
  });
});