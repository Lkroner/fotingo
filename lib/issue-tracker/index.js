'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

var _jira = require('./jira');

var _jira2 = _interopRequireDefault(_jira);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

exports['default'] = _ramda2['default'].always(_jira2['default']);