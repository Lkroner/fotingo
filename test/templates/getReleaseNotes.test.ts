import 'jest';

import { editVirtualFile } from 'src/io/file';
import { Messenger } from 'src/io/messenger';
import { getReleaseNotes } from 'src/templates/getReleaseNotes';
import { data } from 'test/lib/data';

jest.mock('src/io/file', () => ({
  editVirtualFile: jest.fn(),
}));

editVirtualFile as jest.Mock;

describe('getReleaseNotes', () => {
  let messenger: Messenger;
  const onMessage = jest.fn();

  beforeEach(() => {
    messenger = new Messenger();
    messenger.onMessage(onMessage);
  });

  afterEach(() => {
    messenger.stop();
    jest.resetAllMocks();
    jest.clearAllMocks();
  });

  describe('getReleaseNotes', () => {
    test('allows user to edit the release notes', async () => {
      (editVirtualFile as jest.Mock).mockImplementation(
        ({ initialContent }) => `${initialContent} - edited`,
      );
      const releaseNotes = await getReleaseNotes(
        data.createReleaseConfig(),
        messenger,
        data.createRelease(),
        false,
      ).toPromise();
      expect(releaseNotes).toMatchSnapshot();
      expect(onMessage).toHaveBeenCalledTimes(2);
      expect(onMessage.mock.calls[0][0]).toMatchObject({ inThread: true });
      expect(onMessage.mock.calls[1][0]).toMatchObject({ inThread: false });
    });

    test('uses the initial notes if useDefaults is set', async () => {
      const releaseNotes = await getReleaseNotes(
        data.createReleaseConfig(),
        messenger,
        data.createRelease(),
        true,
      ).toPromise();
      expect(releaseNotes).toMatchSnapshot();
      expect(editVirtualFile as jest.Mock).not.toHaveBeenCalled();
      expect(onMessage).not.toHaveBeenCalled();
    });
  });
});
