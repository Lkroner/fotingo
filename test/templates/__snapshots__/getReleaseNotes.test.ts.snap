// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getReleaseNotes getReleaseNotes allows user to edit the release notes 1`] = `
Object {
  "body": "
# Bug fixes:

* [#FOTINGO-2484](https://clark.net): Chief
# Features:

* [#FOTINGO-3908](http://stacey.org): Forward

See [Jira release](http://myrtie.biz)

ðŸš€ Release created with [fotingo](https://github.com/tagoro9/fotingo) - edited",
  "title": "ipsa",
}
`;

exports[`getReleaseNotes getReleaseNotes uses the initial notes if useDefaults is set 1`] = `
Object {
  "body": "
# Bug fixes:

* [#FOTINGO-2484](https://clark.net): Chief
# Features:

* [#FOTINGO-3908](http://stacey.org): Forward

See [Jira release](http://myrtie.biz)

ðŸš€ Release created with [fotingo](https://github.com/tagoro9/fotingo)",
  "title": "ipsa",
}
`;
